// @react-hookz/web/esm/useQueue/useQueue.js
import { useMemo } from 'react';
import { useList } from "../useList/useList.js";
import { useSyncedRef } from "../useSyncedRef/useSyncedRef.js";
/**
 * A state hook implementing FIFO queue.
 *
 * @param initialValue The initial value. Defaults to an empty array.
 */
export function useQueue(initialValue) {
    if (initialValue === void 0) { initialValue = []; }
    var _a = useList(initialValue), list = _a[0], _b = _a[1], removeAt = _b.removeAt, push = _b.push;
    var listRef = useSyncedRef(list);
    return useMemo(function () { return ({
        add: function (value) { return push(value); },
        remove: function () {
            var val = listRef.current[0];
            removeAt(0);
            return val;
        },
        get first() {
            return listRef.current[0];
        },
        get last() {
            return listRef.current[listRef.current.length - 1];
        },
        get size() {
            return listRef.current.length;
        },
        get items() {
            return listRef.current;
        },
    }); }, 
    // eslint-disable-next-line react-hooks/exhaustive-deps
    []);
}
