// @backstage/plugin-tech-radar/dist/alpha.esm.js
import { createApiFactory } from '@backstage/core-plugin-api';
import { createPageExtension, createSchemaFromZod, createApiExtension, createPlugin } from '@backstage/frontend-plugin-api';
import React from 'react';
import { r as rootRouteRef, a as techRadarApiRef, S as SampleTechRadarApi } from './esm/plugin-76dae734.esm.js';
import { convertLegacyRouteRef, compatWrapper } from '@backstage/core-compat-api';

const techRadarPage = createPageExtension({
  defaultPath: "/tech-radar",
  routeRef: convertLegacyRouteRef(rootRouteRef),
  configSchema: createSchemaFromZod(
    (z) => z.object({
      title: z.string().default("Tech Radar"),
      subtitle: z.string().default("Pick the recommended technologies for your projects"),
      pageTitle: z.string().default("Company Radar"),
      path: z.string().default("/tech-radar"),
      width: z.number().default(1500),
      height: z.number().default(800)
    })
  ),
  loader: ({ config }) => import('./esm/index-897dae27.esm.js').then(
    (m) => compatWrapper(/* @__PURE__ */ React.createElement(m.RadarPage, { ...config }))
  )
});
const techRadarApi = createApiExtension({
  factory: createApiFactory(techRadarApiRef, new SampleTechRadarApi())
});
var alpha = createPlugin({
  id: "tech-radar",
  extensions: [techRadarPage, techRadarApi],
  routes: {
    root: convertLegacyRouteRef(rootRouteRef)
  }
});

export { alpha as default, techRadarApi, techRadarPage };
