// @backstage/plugin-events-node/dist/index.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class EventRouter {
  constructor() {
    __publicField(this, "eventBroker");
  }
  async onEvent(params) {
    var _a;
    const topic = this.determineDestinationTopic(params);
    if (!topic) {
      return;
    }
    (_a = this.eventBroker) == null ? void 0 : _a.publish({
      ...params,
      topic
    });
  }
  async setEventBroker(eventBroker) {
    this.eventBroker = eventBroker;
  }
}

class SubTopicEventRouter extends EventRouter {
  constructor(topic) {
    super();
    this.topic = topic;
  }
  determineDestinationTopic(params) {
    const subTopic = this.determineSubTopic(params);
    return subTopic ? `${params.topic}.${subTopic}` : void 0;
  }
  supportsEventTopics() {
    return [this.topic];
  }
}

exports.EventRouter = EventRouter;
exports.SubTopicEventRouter = SubTopicEventRouter;
