// @backstage/plugin-scaffolder-common/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var pluginPermissionCommon = require('@backstage/plugin-permission-common');

const RESOURCE_TYPE_SCAFFOLDER_TEMPLATE = "scaffolder-template";
const RESOURCE_TYPE_SCAFFOLDER_ACTION = "scaffolder-action";
const actionExecutePermission = pluginPermissionCommon.createPermission({
  name: "scaffolder.action.execute",
  attributes: {},
  resourceType: RESOURCE_TYPE_SCAFFOLDER_ACTION
});
const templateParameterReadPermission = pluginPermissionCommon.createPermission({
  name: "scaffolder.template.parameter.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_SCAFFOLDER_TEMPLATE
});
const templateStepReadPermission = pluginPermissionCommon.createPermission({
  name: "scaffolder.template.step.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_SCAFFOLDER_TEMPLATE
});
const scaffolderPermissions = [
  actionExecutePermission,
  templateParameterReadPermission,
  templateStepReadPermission
];
const scaffolderTemplatePermissions = [
  templateParameterReadPermission,
  templateStepReadPermission
];
const scaffolderActionPermissions = [actionExecutePermission];

exports.RESOURCE_TYPE_SCAFFOLDER_ACTION = RESOURCE_TYPE_SCAFFOLDER_ACTION;
exports.RESOURCE_TYPE_SCAFFOLDER_TEMPLATE = RESOURCE_TYPE_SCAFFOLDER_TEMPLATE;
exports.actionExecutePermission = actionExecutePermission;
exports.scaffolderActionPermissions = scaffolderActionPermissions;
exports.scaffolderPermissions = scaffolderPermissions;
exports.scaffolderTemplatePermissions = scaffolderTemplatePermissions;
exports.templateParameterReadPermission = templateParameterReadPermission;
exports.templateStepReadPermission = templateStepReadPermission;
