// @backstage/plugin-scaffolder-common/dist/alpha.esm.js
import { createPermission } from '@backstage/plugin-permission-common';

const RESOURCE_TYPE_SCAFFOLDER_TEMPLATE = "scaffolder-template";
const RESOURCE_TYPE_SCAFFOLDER_ACTION = "scaffolder-action";
const actionExecutePermission = createPermission({
  name: "scaffolder.action.execute",
  attributes: {},
  resourceType: RESOURCE_TYPE_SCAFFOLDER_ACTION
});
const templateParameterReadPermission = createPermission({
  name: "scaffolder.template.parameter.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_SCAFFOLDER_TEMPLATE
});
const templateStepReadPermission = createPermission({
  name: "scaffolder.template.step.read",
  attributes: {
    action: "read"
  },
  resourceType: RESOURCE_TYPE_SCAFFOLDER_TEMPLATE
});
const scaffolderPermissions = [
  actionExecutePermission,
  templateParameterReadPermission,
  templateStepReadPermission
];
const scaffolderTemplatePermissions = [
  templateParameterReadPermission,
  templateStepReadPermission
];
const scaffolderActionPermissions = [actionExecutePermission];

export { RESOURCE_TYPE_SCAFFOLDER_ACTION, RESOURCE_TYPE_SCAFFOLDER_TEMPLATE, actionExecutePermission, scaffolderActionPermissions, scaffolderPermissions, scaffolderTemplatePermissions, templateParameterReadPermission, templateStepReadPermission };
