// @backstage/plugin-catalog-backend-module-azure/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendCommon = require('@backstage/backend-common');
var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha = require('@backstage/plugin-catalog-node/alpha');
var AzureDevOpsEntityProvider = require('./cjs/AzureDevOpsEntityProvider-c37446f5.cjs.js');
require('@backstage/integration');
require('@backstage/plugin-catalog-node');
require('@backstage/backend-tasks');
require('uuid');
require('node-fetch');

const catalogModuleAzureDevOpsEntityProvider = backendPluginApi.createBackendModule({
  pluginId: "catalog",
  moduleId: "azure-dev-ops-entity-provider",
  register(env) {
    env.registerInit({
      deps: {
        config: backendPluginApi.coreServices.rootConfig,
        catalog: alpha.catalogProcessingExtensionPoint,
        logger: backendPluginApi.coreServices.logger,
        scheduler: backendPluginApi.coreServices.scheduler
      },
      async init({ config, catalog, logger, scheduler }) {
        catalog.addEntityProvider(
          AzureDevOpsEntityProvider.AzureDevOpsEntityProvider.fromConfig(config, {
            logger: backendCommon.loggerToWinstonLogger(logger),
            scheduler
          })
        );
      }
    });
  }
});

exports["default"] = catalogModuleAzureDevOpsEntityProvider;
