// @swagger-api/apidom-ns-workflows-1/cjs/refractor/visitors/workflows-1/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _FixedFieldsVisitor = _interopRequireDefault(require("../generics/FixedFieldsVisitor.cjs"));
var _FallbackVisitor = _interopRequireDefault(require("../FallbackVisitor.cjs"));
var _WorkflowsSpecification = _interopRequireDefault(require("../../../elements/WorkflowsSpecification1.cjs"));
const WorkflowsSpecificationVisitor = (0, _stampit.default)(_FixedFieldsVisitor.default, _FallbackVisitor.default, {
  props: {
    specPath: (0, _ramda.always)(['document', 'objects', 'WorkflowsSpecification']),
    canSupportSpecificationExtensions: true
  },
  init() {
    this.element = new _WorkflowsSpecification.default();
  },
  methods: {
    ObjectElement(objectElement) {
      this.unrefractedElement = objectElement;

      // @ts-ignore
      return _FixedFieldsVisitor.default.compose.methods.ObjectElement.call(this, objectElement);
    }
  }
});
var _default = exports.default = WorkflowsSpecificationVisitor;
