// immutable-json-patch/lib/cjs/typeguards.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isJSONArray = isJSONArray;
exports.isJSONObject = isJSONObject;
exports.isJSONPatchAdd = isJSONPatchAdd;
exports.isJSONPatchCopy = isJSONPatchCopy;
exports.isJSONPatchMove = isJSONPatchMove;
exports.isJSONPatchOperation = isJSONPatchOperation;
exports.isJSONPatchRemove = isJSONPatchRemove;
exports.isJSONPatchReplace = isJSONPatchReplace;
exports.isJSONPatchTest = isJSONPatchTest;
function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
function isJSONArray(value) {
  return Array.isArray(value);
}
function isJSONObject(value) {
  return value !== null && _typeof(value) === 'object' && value.constructor === Object // do not match on classes or Array
  ;
}

function isJSONPatchOperation(value) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return value && _typeof(value) === 'object' ? typeof value.op === 'string' : false;
}
function isJSONPatchAdd(value) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return value && _typeof(value) === 'object' ? value.op === 'add' : false;
}
function isJSONPatchRemove(value) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return value && _typeof(value) === 'object' ? value.op === 'remove' : false;
}
function isJSONPatchReplace(value) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return value && _typeof(value) === 'object' ? value.op === 'replace' : false;
}
function isJSONPatchCopy(value) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return value && _typeof(value) === 'object' ? value.op === 'copy' : false;
}
function isJSONPatchMove(value) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return value && _typeof(value) === 'object' ? value.op === 'move' : false;
}
function isJSONPatchTest(value) {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  return value && _typeof(value) === 'object' ? value.op === 'test' : false;
}
