// @backstage/plugin-catalog-backend/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var alpha = require('@backstage/plugin-catalog-common/alpha');
var pluginPermissionNode = require('@backstage/plugin-permission-node');
var CatalogBuilder = require('./cjs/CatalogBuilder-f40ff71c.cjs.js');
var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha$1 = require('@backstage/plugin-catalog-node/alpha');
var backendCommon = require('@backstage/backend-common');
require('@backstage/catalog-model');
require('@backstage/integration');
require('crypto');
require('lodash');
require('@backstage/errors');
require('core-js/features/promise');
require('codeowners-utils');
require('git-url-parse');
require('@backstage/plugin-catalog-node');
require('fs-extra');
require('glob');
require('path');
require('util');
require('yaml');
require('p-limit');
require('uuid');
require('luxon');
require('prom-client');
require('@opentelemetry/api');
require('fast-json-stable-stringify');
require('lodash/uniq');
require('lodash/chunk');
require('zod');
require('@backstage/types');
require('@backstage/catalog-client');
require('yn');
require('@backstage/backend-openapi-utils');
require('@backstage/plugin-auth-node');
require('@backstage/plugin-permission-common');
require('minimatch');
require('@backstage/config');

const { conditions, createConditionalDecision } = pluginPermissionNode.createConditionExports({
  pluginId: "catalog",
  resourceType: alpha.RESOURCE_TYPE_CATALOG_ENTITY,
  rules: CatalogBuilder.permissionRules
});
const catalogConditions = conditions;
const createCatalogConditionalDecision = createConditionalDecision;

var __accessCheck = (obj, member, msg) => {
  if (!member.has(obj))
    throw TypeError("Cannot " + msg);
};
var __privateGet = (obj, member, getter) => {
  __accessCheck(obj, member, "read from private field");
  return getter ? getter.call(obj) : member.get(obj);
};
var __privateAdd = (obj, member, value) => {
  if (member.has(obj))
    throw TypeError("Cannot add the same private member more than once");
  member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
};
var _processors, _entityProviders, _placeholderResolvers, _locationAnalyzers, _permissionRules;
class CatalogProcessingExtensionPointImpl {
  constructor() {
    __privateAdd(this, _processors, new Array());
    __privateAdd(this, _entityProviders, new Array());
    __privateAdd(this, _placeholderResolvers, {});
  }
  addProcessor(...processors) {
    __privateGet(this, _processors).push(...processors.flat());
  }
  addEntityProvider(...providers) {
    __privateGet(this, _entityProviders).push(...providers.flat());
  }
  addPlaceholderResolver(key, resolver) {
    if (key in __privateGet(this, _placeholderResolvers))
      throw new Error(
        `A placeholder resolver for '${key}' has already been set up, please check your config.`
      );
    __privateGet(this, _placeholderResolvers)[key] = resolver;
  }
  get processors() {
    return __privateGet(this, _processors);
  }
  get entityProviders() {
    return __privateGet(this, _entityProviders);
  }
  get placeholderResolvers() {
    return __privateGet(this, _placeholderResolvers);
  }
}
_processors = new WeakMap();
_entityProviders = new WeakMap();
_placeholderResolvers = new WeakMap();
class CatalogAnalysisExtensionPointImpl {
  constructor() {
    __privateAdd(this, _locationAnalyzers, new Array());
  }
  addLocationAnalyzer(analyzer) {
    __privateGet(this, _locationAnalyzers).push(analyzer);
  }
  get locationAnalyzers() {
    return __privateGet(this, _locationAnalyzers);
  }
}
_locationAnalyzers = new WeakMap();
class CatalogPermissionExtensionPointImpl {
  constructor() {
    __privateAdd(this, _permissionRules, new Array());
  }
  addPermissionRules(...rules) {
    __privateGet(this, _permissionRules).push(...rules.flat());
  }
  get permissionRules() {
    return __privateGet(this, _permissionRules);
  }
}
_permissionRules = new WeakMap();
const catalogPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "catalog",
  register(env) {
    const processingExtensions = new CatalogProcessingExtensionPointImpl();
    env.registerExtensionPoint(
      alpha$1.catalogProcessingExtensionPoint,
      processingExtensions
    );
    const analysisExtensions = new CatalogAnalysisExtensionPointImpl();
    env.registerExtensionPoint(
      alpha$1.catalogAnalysisExtensionPoint,
      analysisExtensions
    );
    const permissionExtensions = new CatalogPermissionExtensionPointImpl();
    env.registerExtensionPoint(
      alpha$1.catalogPermissionExtensionPoint,
      permissionExtensions
    );
    env.registerInit({
      deps: {
        logger: backendPluginApi.coreServices.logger,
        config: backendPluginApi.coreServices.rootConfig,
        reader: backendPluginApi.coreServices.urlReader,
        permissions: backendPluginApi.coreServices.permissions,
        database: backendPluginApi.coreServices.database,
        httpRouter: backendPluginApi.coreServices.httpRouter,
        lifecycle: backendPluginApi.coreServices.lifecycle,
        scheduler: backendPluginApi.coreServices.scheduler
      },
      async init({
        logger,
        config,
        reader,
        database,
        permissions,
        httpRouter,
        lifecycle,
        scheduler
      }) {
        const winstonLogger = backendCommon.loggerToWinstonLogger(logger);
        const builder = await CatalogBuilder.CatalogBuilder.create({
          config,
          reader,
          permissions,
          database,
          scheduler,
          logger: winstonLogger
        });
        builder.addProcessor(...processingExtensions.processors);
        builder.addEntityProvider(...processingExtensions.entityProviders);
        Object.entries(processingExtensions.placeholderResolvers).forEach(
          ([key, resolver]) => builder.setPlaceholderResolver(key, resolver)
        );
        builder.addLocationAnalyzers(...analysisExtensions.locationAnalyzers);
        builder.addPermissionRules(...permissionExtensions.permissionRules);
        const { processingEngine, router } = await builder.build();
        await processingEngine.start();
        lifecycle.addShutdownHook(() => processingEngine.stop());
        httpRouter.use(router);
      }
    });
  }
});

exports.createCatalogPermissionRule = CatalogBuilder.createCatalogPermissionRule;
exports.permissionRules = CatalogBuilder.permissionRules;
exports.catalogConditions = catalogConditions;
exports.createCatalogConditionalDecision = createCatalogConditionalDecision;
exports["default"] = catalogPlugin;
