// @backstage/plugin-catalog-backend-module-github/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var backendCommon = require('@backstage/backend-common');
var alpha = require('@backstage/plugin-catalog-node/alpha');
var GithubEntityProvider = require('./cjs/GithubEntityProvider-858e0576.cjs.js');
require('@backstage/integration');
require('@backstage/plugin-catalog-node');
require('@octokit/graphql');
require('uuid');
require('@backstage/backend-tasks');
require('@backstage/catalog-model');
require('lodash');
require('minimatch');

const catalogModuleGithubEntityProvider = backendPluginApi.createBackendModule({
  pluginId: "catalog",
  moduleId: "github-entity-provider",
  register(env) {
    env.registerInit({
      deps: {
        catalog: alpha.catalogProcessingExtensionPoint,
        config: backendPluginApi.coreServices.rootConfig,
        logger: backendPluginApi.coreServices.logger,
        scheduler: backendPluginApi.coreServices.scheduler
      },
      async init({ catalog, config, logger, scheduler }) {
        catalog.addEntityProvider(
          GithubEntityProvider.GithubEntityProvider.fromConfig(config, {
            logger: backendCommon.loggerToWinstonLogger(logger),
            scheduler
          })
        );
      }
    });
  }
});

exports["default"] = catalogModuleGithubEntityProvider;
