// @backstage/plugin-permission-backend/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendCommon = require('@backstage/backend-common');
var backendPluginApi = require('@backstage/backend-plugin-api');
var alpha = require('@backstage/plugin-permission-node/alpha');
var router = require('./cjs/router-40c490a9.cjs.js');
require('zod');
require('express');
require('express-promise-router');
require('@backstage/errors');
require('@backstage/plugin-permission-common');
require('node-fetch');
require('lodash');
require('dataloader');

var __defProp = Object.defineProperty;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __publicField = (obj, key, value) => {
  __defNormalProp(obj, typeof key !== "symbol" ? key + "" : key, value);
  return value;
};
class PolicyExtensionPointImpl {
  constructor() {
    __publicField(this, "policy");
  }
  setPolicy(policy) {
    if (this.policy) {
      throw new Error("Policy already set");
    }
    this.policy = policy;
  }
}
const permissionPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "permission",
  register(env) {
    const policies = new PolicyExtensionPointImpl();
    env.registerExtensionPoint(alpha.policyExtensionPoint, policies);
    env.registerInit({
      deps: {
        http: backendPluginApi.coreServices.httpRouter,
        config: backendPluginApi.coreServices.rootConfig,
        logger: backendPluginApi.coreServices.logger,
        discovery: backendPluginApi.coreServices.discovery,
        identity: backendPluginApi.coreServices.identity
      },
      async init({ http, config, logger, discovery, identity }) {
        const winstonLogger = backendCommon.loggerToWinstonLogger(logger);
        if (!policies.policy) {
          throw new Error(
            "No policy module installed! Please install a policy module. If you want to allow all requests, use @backstage/plugin-permission-backend-module-allow-all-policy permissionModuleAllowAllPolicy"
          );
        }
        http.use(
          await router.createRouter({
            config,
            discovery,
            identity,
            logger: winstonLogger,
            policy: policies.policy
          })
        );
      }
    });
  }
});

exports["default"] = permissionPlugin;
