// @backstage/plugin-search/dist/esm/index-eb06eaf0.esm.js
import React from 'react';
import SearchIcon from '@material-ui/icons/Search';
import { SidebarItem } from '@backstage/core-components';
import { a as SearchModalProvider, u as useSearchModal, S as SearchModal } from './useSearchModal-77675f70.esm.js';
import '@backstage/core-plugin-api';
import '@backstage/plugin-search-react';
import '@material-ui/core';
import '@material-ui/core/Box';
import '@material-ui/core/Button';
import '@material-ui/core/IconButton';
import '@material-ui/core/styles';
import '@material-ui/icons/ArrowForward';
import '@material-ui/icons/Close';
import 'react-router-dom';
import './SearchType-a8f79021.esm.js';
import '@backstage/errors';
import 'qs';
import 'react-use/lib/usePrevious';
import 'react-use/lib/useEffectOnce';
import '@material-ui/icons/ExpandMore';
import '@material-ui/icons/FontDownload';
import 'react-use/lib/useAsync';
import '@backstage/version-bridge';
import 'react-use/lib/useUpdateEffect';

const SidebarSearchModalContent = (props) => {
  const { state, toggleModal } = useSearchModal();
  const Icon = props.icon ? props.icon : SearchIcon;
  return /* @__PURE__ */ React.createElement(React.Fragment, null, /* @__PURE__ */ React.createElement(
    SidebarItem,
    {
      className: "search-icon",
      icon: Icon,
      text: "Search",
      onClick: toggleModal
    }
  ), /* @__PURE__ */ React.createElement(
    SearchModal,
    {
      ...state,
      toggleModal,
      children: props.children
    }
  ));
};
const SidebarSearchModal = (props) => {
  return /* @__PURE__ */ React.createElement(SearchModalProvider, null, /* @__PURE__ */ React.createElement(SidebarSearchModalContent, { ...props }));
};

export { SidebarSearchModal };
